<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Pyomic.single.cosg &mdash; Pyomic 1.0.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Pyomic.single.data_preprocess" href="Pyomic.single.data_preprocess.html" />
    <link rel="prev" title="Pyomic.bulk.string_map" href="Pyomic.bulk.string_map.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Pyomic
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Installation_guild.html">Installation guild</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Tutorials/Tutorials.html">Tutorial</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../API_documentation.html">API documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../API_documentation.html#bulk">bulk</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../API_documentation.html#single">single</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Pyomic.single.cosg</a></li>
<li class="toctree-l3"><a class="reference internal" href="Pyomic.single.data_preprocess.html">Pyomic.single.data_preprocess</a></li>
<li class="toctree-l3"><a class="reference internal" href="Pyomic.single.cell_annotate.html">Pyomic.single.cell_annotate</a></li>
<li class="toctree-l3"><a class="reference internal" href="Pyomic.single.cell_anno_print.html">Pyomic.single.cell_anno_print</a></li>
<li class="toctree-l3"><a class="reference internal" href="Pyomic.single.scnocd.html">Pyomic.single.scnocd</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../Example_dataset.html">Example dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Release_notes.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Developer_guild.html">Developer guild</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Pyomic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../API_documentation.html">API documentation</a> &raquo;</li>
      <li>Pyomic.single.cosg</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/Pyomic.single.cosg.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="pyomic-single-cosg">
<h1>Pyomic.single.cosg<a class="headerlink" href="#pyomic-single-cosg" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="Pyomic.single.cosg">
<span class="sig-prename descclassname"><span class="pre">Pyomic.single.</span></span><span class="sig-name descname"><span class="pre">cosg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CellTypes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_lowly_expressed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expressed_pct</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_genes_user</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calculate_logfoldchanges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/Pyomic/single/_cosg.html#cosg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Pyomic.single.cosg" title="Permalink to this definition"></a></dt>
<dd><p>Marker gene identification for single-cell sequencing data using COSG.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Annotated data matrix. Note: input paramters are simliar to the parameters used for scanpy’s rank_genes_groups() function.</p></li>
<li><p><strong>groupby</strong> – The key of the cell groups in .obs, the default value is set to ‘CellTypes’.</p></li>
<li><p><strong>groups</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">typing.Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Literal" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">typing.Literal</span></code></a>[‘all’], <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterable" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">typing.Iterable</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – Subset of cell groups, e.g. [<cite>‘g1’</cite>, <cite>‘g2’</cite>, <cite>‘g3’</cite>], to which comparison shall be restricted. The default value is ‘all’, and all groups will be compared.</p></li>
<li><p><strong>mu</strong> – The penalty restricting marker genes expressing in non-target cell groups. Larger value represents more strict restrictions. mu should be &gt;= 0, and by default, mu = 1.</p></li>
<li><p><strong>remove_lowly_expressed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – If True, genes that express a percentage of target cells smaller than a specific value (<cite>expressed_pct</cite>) are not considered as marker genes for the target cells. The default value is False.</p></li>
<li><p><strong>expressed_pct</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">typing.Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]) – When <cite>remove_lowly_expressed</cite> is set to True, genes that express a percentage of target cells smaller than a specific value (<cite>expressed_pct</cite>) are not considered as marker genes for the target cells. The default value for <cite>expressed_pct</cite>
is 0.1 (10%).</p></li>
<li><p><strong>n_genes_user</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – The number of genes that appear in the returned tables. The default value is 50.</p></li>
<li><p><strong>key_added</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">typing.Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – The key in <cite>adata.uns</cite> information is saved to.</p></li>
<li><p><strong>calculate_logfoldchanges</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Calculate logfoldchanges.</p></li>
<li><p><strong>use_raw</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Use <cite>raw</cite> attribute of <cite>adata</cite> if present.</p></li>
<li><p><strong>layer</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">typing.Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – Key from <cite>adata.layers</cite> whose value will be used to perform tests on.</p></li>
<li><p><strong>reference</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – If <cite>‘rest’</cite>, compare each group to the union of the rest of the group.
If a group identifier, compare with respect to this group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>names</strong> (structured <cite>np.ndarray</cite> (<cite>.uns[‘rank_genes_groups’]</cite>)) – Structured array to be indexed by group id storing the gene names. Ordered according to scores.</p></li>
<li><p><strong>scores</strong> (structured <cite>np.ndarray</cite> (<cite>.uns[‘rank_genes_groups’]</cite>)) – Structured array to be indexed by group id storing COSG scores for each gene for each
group. Ordered according to scores.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Contact: Min Dai, <a class="reference external" href="mailto:daimin&#37;&#52;&#48;zju&#46;edu&#46;cn">daimin<span>&#64;</span>zju<span>&#46;</span>edu<span>&#46;</span>cn</a></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">cosg</span> <span class="k">as</span> <span class="nn">cosg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc68k_reduced</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cosg</span><span class="o">.</span><span class="n">cosg</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;cosg&#39;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;bulk_labels&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s1">&#39;cosg&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Pyomic.bulk.string_map.html" class="btn btn-neutral float-left" title="Pyomic.bulk.string_map" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Pyomic.single.data_preprocess.html" class="btn btn-neutral float-right" title="Pyomic.single.data_preprocess" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, starlitnightly.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>